File: starfightersquadrons/src/main/java/org/vizzini/starfightersquadrons/Weapon.java
package org.vizzini.starfightersquadrons;

import java.util.Set;

import org.vizzini.starfightersquadrons.RangeRuler.Range;

/**
 * Defines methods required by a weapon.
 */
public interface Weapon
{
    /**
     * @return the name
     */
    String getName();

    /**
     * @return the ranges
     */
    Set<Range> getRanges();

    /**
     * @return the weaponValue
     */
    int getWeaponValue();

    /**
     * @param attacker Attacker.
     * @param attackerPosition Attacker position.
     * @param defender Defender.
     * @param defenderPosition Defender position.
     * 
     * @return true if the defender is in this weapon's range.
     */
    boolean isDefenderInRange(final SSToken attacker, final SSPosition attackerPosition,
            final SSToken defender, final SSPosition defenderPosition);

    /**
     * @param attacker Attacker.
     * @param attackerPosition Attacker position.
     * @param defender Defender.
     * @param defenderPosition Defender position.
     * 
     * @return true if the defender is in this weapon's firing arc and range.
     */
    boolean isDefenderTargetable(final SSToken attacker, final SSPosition attackerPosition,
            final SSToken defender, final SSPosition defenderPosition);

    /**
     * @param attacker Attacker.
     * @param attackerPosition Attacker position.
     * @param defender Defender.
     * @param defenderPosition Defender position.
     * 
     * @return true if the defender is in this weapon's firing arc.
     */
    boolean isDefenderVulnerable(final SSToken attacker, final SSPosition attackerPosition,
            final SSToken defender, final SSPosition defenderPosition);

    /**
     * @return the isPrimary
     */
    boolean isPrimary();
}

File: starfightersquadronsweb/src/main/js/Weapon.js
define([ "FiringArc", "ManeuverComputer", "RangeRuler", "UpgradeCard", "UpgradeHeader" ], function(FiringArc,
        ManeuverComputer, RangeRuler, UpgradeCard, UpgradeHeader)
{
    "use strict";
    function Weapon(name, weaponValue, ranges, firingArcKey, upgradeKey)
    {
        InputValidator.validateNotNull("name", name);
        InputValidator.validateNotNull("weaponValue", weaponValue);
        InputValidator.validateNotNull("ranges", ranges);
        InputValidator.validateNotNull("firingArcKey", firingArcKey);
        // upgradeKey optional.

        this.name = function()
        {
            return name;
        };

        this.weaponValue = function()
        {
            return weaponValue;
        };

        this.ranges = function()
        {
            return ranges;
        };

        this.firingArcKey = function()
        {
            return firingArcKey;
        };

        this.upgradeKey = function()
        {
            return upgradeKey;
        };

        var upgrade = UpgradeCard.properties[upgradeKey];

        this.upgrade = function()
        {
            return upgrade;
        };
    }

    Weapon.prototype.isUsable = function(attacker, defender)
    {
        var answer;
        var upgrade = this.upgrade();

        if (upgrade)
        {
            if (upgrade.isImplemented)
            {
                switch (upgrade.header)
                {
                case UpgradeHeader.ATTACK:
                    answer = true;
                    break;
                case UpgradeHeader.ATTACK_FOCUS:
                    answer = (attacker.focus().count() > 0);
                    break;
                case UpgradeHeader.ATTACK_TARGET_LOCK:
                    answer = (attacker.findTargetLockByDefender(defender) !== undefined);
                    break;
                default:
                    throw "Unknown upgrade header: " + header;
                }
            }
            else
            {
                answer = false;
            }
        }
        else
        {
            // Primary weapon.
            answer = true;
        }

        return answer;
    };

    Weapon.prototype.isDefenderInFiringArc = function(attackerPosition, defender, defenderPosition)
    {
        InputValidator.validateNotNull("attackerPosition", attackerPosition);
        InputValidator.validateNotNull("defender", defender);
        InputValidator.validateNotNull("defenderPosition", defenderPosition);

        var firingArc = FiringArc.properties[this.firingArcKey()];
        var bearing = attackerPosition.computeBearing(defenderPosition.x(), defenderPosition.y());
        var answer = firingArc.isInFiringArc(bearing);
        LOGGER.debug("weapon = " + this.name());
        LOGGER.debug("0 firingArcKey = " + this.firingArcKey() + " bearing = " + bearing + " answer ? " + answer);

        if (!answer)
        {
            var shipBaseKey = defender.pilot().shipTeam.ship.shipBaseKey;
            var polygon = ManeuverComputer.computePolygon(shipBaseKey, defenderPosition.x(), defenderPosition.y(),
                    defenderPosition.heading());
            var points = polygon.points();

            for (var i = 0; i < points.length; i += 2)
            {
                bearing = attackerPosition.computeBearing(points[i], points[i + 1]);

                if (firingArc.isInFiringArc(bearing))
                {
                    answer = true;
                    LOGGER.debug(i + " firingArcKey = " + this.firingArcKey() + " bearing = " + bearing + " answer ? " +
                            answer);
                    break;
                }
            }
        }

        return answer;
    };

    Weapon.prototype.isDefenderInRange = function(attacker, attackerPosition, defender, defenderPosition)
    {
        InputValidator.validateNotNull("attacker", attacker);
        InputValidator.validateNotNull("attackerPosition", attackerPosition);
        InputValidator.validateNotNull("defender", defender);
        InputValidator.validateNotNull("defenderPosition", defenderPosition);

        var range = RangeRuler.getRange(attacker, attackerPosition, defender, defenderPosition);

        return this.ranges().vizziniContains(range);
    };

    Weapon.prototype.isDefenderTargetable = function(attacker, attackerPosition, defender, defenderPosition)
    {
        InputValidator.validateNotNull("attacker", attacker);
        InputValidator.validateNotNull("attackerPosition", attackerPosition);
        InputValidator.validateNotNull("defender", defender);
        InputValidator.validateNotNull("defenderPosition", defenderPosition);

        return this.isUsable(attacker, defender) &&
                this.isDefenderInRange(attacker, attackerPosition, defender, defenderPosition) &&
                this.isDefenderInFiringArc(attackerPosition, defender, defenderPosition);
    };

    Weapon.prototype.isPrimary = function()
    {
        return this.name() === "Primary Weapon";
    };

    Weapon.prototype.toString = function()
    {
        return this.name();
    };

    return Weapon;
});
